src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisResult.cs

```diff
@@ -0,0 +1,208 @@
+        private readonly Dictionary<DiagnosticAnalyzer, TimeSpan> _analyzerExecutionTimeOpt;
+
+        private Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, List<Diagnostic>>> _localSemanticDiagnosticsOpt = null;
+        private Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, List<Diagnostic>>> _localSyntaxDiagnosticsOpt = null;
+        private Dictionary<DiagnosticAnalyzer, List<Diagnostic>> _nonLocalDiagnosticsOpt = null;
+        
+        public AnalysisResult(bool logAnalyzerExecutionTime, ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            _analyzerExecutionTimeOpt = logAnalyzerExecutionTime ? CreateAnalyzerExecutionTimeMap(analyzers) : null;
+        }
+
+        private static Dictionary<DiagnosticAnalyzer, TimeSpan> CreateAnalyzerExecutionTimeMap(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            var map = new Dictionary<DiagnosticAnalyzer, TimeSpan>(analyzers.Length);
+            foreach (var analyzer in analyzers)
+            {
+                map[analyzer] = default(TimeSpan);
+            }
+
+            return map;
+        }
+
+        public TimeSpan GetAnalyzerExecutionTime(DiagnosticAnalyzer analyzer)
+        {
+            Debug.Assert(_analyzerExecutionTimeOpt != null);
+
+            lock (_gate)
+            {
+                return _analyzerExecutionTimeOpt[analyzer];
+            }
+        }
+
+        public void Update(AnalysisScope analysisScope, AnalyzerDriver driver)
+        {
+            foreach (var analyzer in analysisScope.Analyzers)
+            {
+                // Get reported diagnostics and add them to our local diagnostic caches.
+                var syntaxDiagnostics = driver.GetLocalDiagnostics(analyzer, syntax: true);
+                var semanticDiagnostics = driver.GetLocalDiagnostics(analyzer, syntax: false);
+                var compilationDiagnostics = driver.GetNonLocalDiagnostics(analyzer);
+
+                lock (_gate)
+                {
+                    if (syntaxDiagnostics.Length > 0 || semanticDiagnostics.Length > 0 || compilationDiagnostics.Length > 0)
+                    {
+                        UpdateLocalDiagnosticsCache_NoLock(analyzer, syntaxDiagnostics, ref _localSyntaxDiagnosticsOpt);
+                        UpdateLocalDiagnosticsCache_NoLock(analyzer, semanticDiagnostics, ref _localSemanticDiagnosticsOpt);
+                        UpdateNonLocalDiagnosticsCache_NoLock(analyzer, compilationDiagnostics);
+                    }
+
+                    if (_analyzerExecutionTimeOpt != null)
+                    {
+                        _analyzerExecutionTimeOpt[analyzer] += driver.ResetAnalyzerExecutionTime(analyzer);
+                    }
+                }
+            }
+        }
+
+        private void UpdateLocalDiagnosticsCache_NoLock(DiagnosticAnalyzer analyzer, ImmutableArray<Diagnostic> diagnostics, ref Dictionary<SyntaxTree, Dictionary<DiagnosticAnalyzer, List<Diagnostic>>> lazyLocalDiagnostics)
```

Comment:
We should be careful to use the word "cache" only for caches, where a cache is strictly a performance enhancement and is allowed to discard values.